[SUREFIRE-2011] Updated abstractions which helps associating standard out/err with a test

Tibor17 opened the pull request:
The method `void writeTestOutput( String output, boolean newLine, boolean stdout )` appeared in `ForkingRunListener` and `TestSetRunListener` and it was called by `ConsoleOutputCapture`. 

The information (`testRunId` and Thread) is associated with the particular run of the test method in the implementation of provider's `RunListener`. So the listener has this information, and not the `ForkingRunListener`.
Due to we will implement enum `RunMode` and `testRunId:long` we should not redirect `void writeTestOutput( String output, boolean newLine, boolean stdout )` to `ForkingRunListener`. Therefore we created the interface `TestRunListener` which is implemented by the provider's listener.

After the previous refactoring of surefire-junit3, we should continue with updating the abstraction in order to complete [SUREFIRE-1860](https://issues.apache.org/jira/browse/SUREFIRE-1860). The changes in SUREFIRE-1860 are big and therefore I would like to split them to an abstraction in this PR, continue with another PRs regarding implementation of encoder/decoder, SimpleReportEntry. It would give us the opportunity to associate the std/out/err logs with test run id (Thread) and deterministically create the reports and this way fix pending issues (junit5, and simplify the listeners in junit4.7 provider). So I am splitting the work in several pieces.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SUREFIRE) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SUREFIRE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `SUREFIRE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [ ] You have run the integration tests successfully (`mvn -Prun-its clean install`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

slawekjaranowski commented on surefire-api/src/main/java/org/apache/maven/surefire/api/report/TestRunListener.java:
```diff
@@ -0,0 +1,28 @@
+ * under the License.
+ */
+
+/**
+ *
+ */
+public interface TestRunListener
+    extends RunListener, TestOutputReceiver
+{
+}
```
New interface -  it will be good to add javadoc with description what purpose of this interface is.

Tibor17 replied:
done

slawekjaranowski commented on surefire-providers/surefire-junit47/src/main/java/org/apache/maven/surefire/junitcore/ConcurrentRunListener.java:
```diff
@@ -61,10 +60,10 @@
-        reporterManagerThreadLocal = new ThreadLocal<RunListener>()
+        reporterManagerThreadLocal = new ThreadLocal<TestRunListener>()
         {
             @Override
-            protected RunListener initialValue()
+            protected TestRunListener initialValue()
             {
                 return reporterFactory.createReporter();
             }
```
now can be like:
```
 reporterManagerThreadLocal = ThreadLocal.withInitial( reporterFactory::createReporter );

```

TestSetFailedException is not thrown from this constructor - line 58

Tibor17 replied:
done

slawekjaranowski commented on surefire-providers/common-junit4/src/main/java/org/apache/maven/surefire/common/junit4/JUnit4RunListener.java:
```diff
@@ -41,12 +43,14 @@
+    implements TestOutputReceiver
 {
-    protected final RunListener reporter;
+    protected final TestRunListener reporter;
```
maybe also change reported -> testRunListener
and in other similar place

Tibor17 replied:
The interface `TestReportListener` contains the wording `report`, pls see the next commit.

slawekjaranowski commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/TestSetRunListener.java:
```diff
@@ -51,13 +50,13 @@
  */
 public class TestSetRunListener
-    implements RunListener, ConsoleOutputReceiver, ConsoleLogger
+    implements TestRunListener, ConsoleLogger
```
still we need implements `ConsoleLogger`?
class `TestSetRunListener` is created in one place in `DefaultReporterFactory#createReporter` as  `TestRunListener` ...

slawekjaranowski replied:
Ok, I can wrong

On fork side 
`ForkingRunListener` is used as ConsoleLogger
 instance of `ForkingRunListener` is created by `ForkingReporterFactory#createReporter` 

in `ForkedBooter` I see
```
        forkingReporterFactory = createForkingReporterFactory();
        logger = (ConsoleLogger) forkingReporterFactory.createReporter();
```

but when I see `ConsoleLoggerDecorator` - everything can happen in runtime
so don't spend more time on this ... 
maybe after more time it will be more clear for me

Tibor17 replied:
The data flow `fork` -> `plugin` is all about reporting events.
The data flow `plugin` -> `fork` is about commands.
So this is Command Event based architecture. Similar wordings are in the enterprise world CQRS and Event Sourcing.

The first data flow usually sends objects in the form of:
- new XxxEvent(new ReportEntry()) - typically test success
- new XxxEvent(<string>) - typically std/out from a Thread of a running test
- new XxxEvent(<string>) - typically console error, debug

Therefore there is such a wording `report` because the provider (InPlugin or fork) sends a report back to the plugin, and the provider is under the plugin in the hierarchy because the plugin is a manager of providers.

Now it is visible in the inheritance
`ForkingRunListener` -> `RunListener`, `TestOutputReceiver`, `ConsoleLogger`.
These are 3 types and three interfaces.

I know that you have a problem with the name of class `ForkingRunListener` but it is not mandatory for the developers because the developers and their code operates with abstractions and their code instantiates the object somewhere but the interfaces are use everywhere. The interfaces have different purpose because we do not want `ParallelComputer` to send TEST_SUCCESS of course not, and the PC sends console internal error if any. This is called information hiding. **So there are reports, all these three interfaces are about reports but different group of reports.**

Regarding `ForkingRunListener`, whatsoever do you like `ForkingReportListener` more?

From the history, you should know how the code was before. These methods of reports were everywhere. They did not have one place like it is not. Their implementation was really everywhere, the encoder did not exist in one place and so it was very hard to maintain the code. Naturally this situation happened because we are OSS, and in the OSS you start from simple use cases and you do not have time to think of encapsulation, information hiding, and such things.

Tibor17 replied:
@slawekjaranowski 
This line of code `logger = (ConsoleLogger) forkingReporterFactory.createReporter();` does not exist in the second commit.

Tibor17 replied:
Currently, we have to create a new instance of the listener via `createReporter()` because the Provider `surefire-junit47` has own instance in ThreadLocal. After this has finished in [SUREFIRE-1643](https://issues.apache.org/jira/browse/SUREFIRE-1643) and [SUREFIRE-1860](https://issues.apache.org/jira/browse/SUREFIRE-1860), we can refactor `surefire-junit47` where the ThreadLocal would not be needed and the creator of reporter may turn to getter of singleton.

Tibor17 replied:
@slawekjaranowski 
If you mean to use `ConsoleLoggerDecorator` in order to cast `Object` to `ConsoleLogger`, then I have to say that it is too heavy procedure. Now the second commit does not use casting types because we have a composition of interfaces, so we return the subinterface.

slawekjaranowski commented on surefire-api/src/main/java/org/apache/maven/surefire/api/report/ReporterFactory.java:
```diff
@@ -33,7 +33,7 @@
      * @return A reporter instance
      */
-    RunListener createReporter();
+    TestRunListener createReporter();
```
rename method `createReporter` -> `crateTestRunListener` WDYT?

javadoc for this method is still actual?

Tibor17 replied:
In this architecture we have two sides (forked JVM, plugin JVM) having implemented the same interface RunListener:
1. side is ForkingRunListener in the fork JVM
2. TestSetRunListener in the plugin JVM

The serialization between these two JVM ensures that the ReportEntries are transfered from the fork to the plugin.
This is the whole idea behind them.

We can see these renaming issues from the another angle, and so that there is a reporter. Listener means that our surefire `ForkingRunListener` is a delegate within JUnit's RunListeners in the fork. On the other side, the TestSetRunListener is a listener registered in the `ForkClient`.

We can think about it a bit deeper and we will find the best names for these interfaces and methods.
Anyway, we will remove deprecated code, we will create Builders for constructors with too many parameters, so we will have a chance to refactor a lot.

Tibor17 replied:
@slawekjaranowski yes, the method was renamed in the second commit.

Tibor17 commented:
@slawekjaranowski 
I have made one more commit because both listeners (ForkingRunListener and TestSetRunListener) are a mirror of the same fork/plugin JVM and so they should implement one and the same interface as mandatory. This way we have avoided ugly casting to type. Any combinations of two interfaces out of three (RunListener, TestOutputReceiver, ConsoleLogger) are avoided and the only TestReportListener is used. Removed ConsoleStream interface and used ConsoleLogger instead. Simplified code around logger in JUnitCoreProvider. Renamed method in ReporterFactory.

Tibor17 commented:
@slawekjaranowski 
Let's have a look. That's all from my side.

slawekjaranowski commented on surefire-providers/surefire-junit47/src/test/java/org/apache/maven/surefire/junitcore/Surefire746Test.java:
```diff
@@ -36,26 +36,17 @@
 import org.apache.maven.surefire.api.booter.ProviderParameterNames;
-import org.apache.maven.surefire.common.junit4.JUnit4RunListener;
-import org.apache.maven.surefire.common.junit4.Notifier;
-import org.apache.maven.surefire.api.report.DefaultDirectConsoleReporter;
 import org.apache.maven.surefire.api.report.ReporterConfiguration;
 import org.apache.maven.surefire.api.report.ReporterFactory;
-import org.apache.maven.surefire.api.report.RunListener;
+import org.apache.maven.surefire.api.report.TestReportListener;
 import org.apache.maven.surefire.api.suite.RunResult;
 import org.apache.maven.surefire.api.testset.TestSetFailedException;
 import org.apache.maven.surefire.api.util.TestsToRun;
-
+import org.apache.maven.surefire.common.junit4.JUnit4RunListener;
+import org.apache.maven.surefire.common.junit4.Notifier;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
```
invalid import order

Tibor17 replied:
fixed

Tibor17 commented on surefire-providers/surefire-junit47/src/test/java/org/apache/maven/surefire/junitcore/ConcurrentRunListenerTest.java:
```diff
@@ -19,26 +19,24 @@
-import org.apache.maven.surefire.report.RunStatistics;
-import org.apache.maven.surefire.api.testset.TestSetFailedException;
-
 import junit.framework.Assert;
 import junit.framework.TestCase;
 import junit.framework.TestSuite;
+import org.apache.maven.plugin.surefire.report.DefaultReporterFactory;
+import org.apache.maven.surefire.api.report.ReporterFactory;
+import org.apache.maven.surefire.api.report.TestReportListener;
+import org.apache.maven.surefire.api.testset.TestSetFailedException;
+import org.apache.maven.surefire.report.RunStatistics;
 import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.runner.Computer;
 import org.junit.runner.JUnitCore;
 
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.util.HashMap;
+import java.util.Map;
+
```
@slawekjaranowski 
Yesterday I downloaded the latest `maven-idea-codestyle.xml` and used in my IDEA.
Can you pls use the same and reproduce this problem?
Try to press CTRL+ALT+O on this file in e.g. master.

Tibor17 replied:
fixed

Tibor17 commented on surefire-providers/surefire-junit47/pom.xml:
```diff
@@ -73,100 +79,6 @@
-                <configuration>
-                    <compilerArguments>
-                        <endorseddirs>${project.build.directory}/endorsed</endorseddirs>
-                    </compilerArguments>
-                    <testCompilerArguments>
-                        <endorseddirs>${project.build.directory}/endorsed-test</endorseddirs>
-                    </testCompilerArguments>
-                </configuration>
-            </plugin>
             <plugin>
```
Do you know the annotation `ThreadSafe` in JUnit4?
Threadsafe listeners.
This can be applied only on JUnit's `RunListener`.

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/booterclient/output/ForkClient.java:
```diff
@@ -423,14 +424,14 @@ public final RunListener getReporter()
     }
 
     private ConsoleLogger getOrCreateConsoleLogger()
     {
-        return (ConsoleLogger) getTestSetReporter();
+        return getTestSetReporter();
     }
```
Because this PR has only limited refactoring. This PR is aming for refactoring of the interfaces of the listeners and the reason is the fact that we are not able to associate system output printed during the test without mapping the log message and test Thread. Other refactoring is out of the scope.

Tibor17 commented on surefire-booter/src/test/java/org/apache/maven/surefire/booter/SurefireReflectorTest.java:
```diff
@@ -55,7 +55,7 @@ public void testShouldCreateFactoryWithoutException()
         ReporterFactory factory = new ReporterFactory()
```
Because this PR has only limited scope of refactoring. This PR is aming for refactoring of the interfaces of the listeners and the reason is the fact that we are not able to associate system output with the Thread of the test without refactoring of these interfaces.

Tibor17 commented on surefire-booter/src/test/java/org/apache/maven/surefire/booter/SurefireReflectorTest.java:
```diff
@@ -255,7 +255,7 @@ public void testReporterFactory()
         ReporterFactory reporterFactory = new ReporterFactory()
```
Because this PR has only limited scope of refactoring. This PR is aming for refactoring of the interfaces of the listeners and the reason is the fact that we are not able to associate system output with the Thread of the test without refactoring of these interfaces.

Tibor17 commented on surefire-booter/src/test/java/org/apache/maven/surefire/booter/SurefireReflectorTest.java:
```diff
@@ -279,7 +279,7 @@ public void testReporterFactoryAware()
         ReporterFactory reporterFactory = new ReporterFactory()
```
@slawekjaranowski This PR has only limited scope of refactoring. This PR is aming for refactoring of the interfaces of the listeners and the reason is the fact that we are not able to associate system output with the Thread of the test without refactoring of these interfaces.

Tibor17 commented on surefire-providers/common-junit4/src/main/java/org/apache/maven/surefire/common/junit4/JUnit4RunListener.java:
```diff
@@ -58,11 +63,16 @@
         this.reporter = reporter;
     }
 
+    public final ConsoleLogger getConsoleLogger()
```
There cannot be Override because there is no such super type having a method to override.
This class in another module, so it must be public and of course without Overide - no abstract method in super type.
